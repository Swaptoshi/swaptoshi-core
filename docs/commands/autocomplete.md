# `swaptoshi-core autocomplete`

display autocomplete installation instructions

- [`swaptoshi-core autocomplete [SHELL]`](#swaptoshi-core-autocomplete-shell)

## `swaptoshi-core autocomplete [SHELL]`

display autocomplete installation instructions

```
USAGE
  $ swaptoshi-core autocomplete [SHELL] [-r]

ARGUMENTS
  SHELL  shell type

FLAGS
  -r, --refresh-cache  Refresh cache (ignores displaying instructions)

DESCRIPTION
  display autocomplete installation instructions

EXAMPLES
  $ swaptoshi-core autocomplete

  $ swaptoshi-core autocomplete bash

  $ swaptoshi-core autocomplete zsh

  $ swaptoshi-core autocomplete --refresh-cache
```

_See code: [@oclif/plugin-autocomplete](https://github.com/oclif/plugin-autocomplete/blob/v1.3.6/src/commands/autocomplete/index.ts)_
