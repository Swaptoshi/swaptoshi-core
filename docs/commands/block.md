# `swaptoshi-core block`

Commands relating to swaptoshi-core blocks.

- [`swaptoshi-core block get INPUT`](#swaptoshi-core-block-get-input)

## `swaptoshi-core block get INPUT`

Get block information for a given id or height.

```
USAGE
  $ swaptoshi-core block get [INPUT] [-d <value>] [--pretty]

ARGUMENTS
  INPUT  Height in number or block id in hex format.

FLAGS
  -d, --data-path=<value>  Directory path to specify where node data is stored. Environment variable "KLAYR_DATA_PATH"
                           can also be used.
  --pretty                 Prints JSON in pretty format rather than condensed.

DESCRIPTION
  Get block information for a given id or height.

EXAMPLES
  block:get e082e79d01016632c451c9df9276e486cb7f460dc793ff5b10d8f71eecec28b4

  block:get 2
```
