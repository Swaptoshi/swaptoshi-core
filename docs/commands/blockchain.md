# `swaptoshi-core blockchain`

Commands relating to swaptoshi-core blockchain data.

- [`swaptoshi-core blockchain export`](#swaptoshi-core-blockchain-export)
- [`swaptoshi-core blockchain hash`](#swaptoshi-core-blockchain-hash)
- [`swaptoshi-core blockchain import FILEPATH`](#swaptoshi-core-blockchain-import-filepath)
- [`swaptoshi-core blockchain reset`](#swaptoshi-core-blockchain-reset)

## `swaptoshi-core blockchain export`

Export to <FILE>.

```
USAGE
  $ swaptoshi-core blockchain export [-d <value>] [-o <value>]

FLAGS
  -d, --data-path=<value>  Directory path to specify where node data is stored. Environment variable "KLAYR_DATA_PATH"
                           can also be used.
  -o, --output=<value>     The output directory. Default will set to current working directory.

DESCRIPTION
  Export to <FILE>.

EXAMPLES
  blockchain:export

  blockchain:export --data-path ./data --output ./my/path/
```

## `swaptoshi-core blockchain hash`

Generate SHA256 hash from <PATH>.

```
USAGE
  $ swaptoshi-core blockchain hash [-d <value>]

FLAGS
  -d, --data-path=<value>  Directory path to specify where node data is stored. Environment variable "KLAYR_DATA_PATH"
                           can also be used.

DESCRIPTION
  Generate SHA256 hash from <PATH>.

EXAMPLES
  blockchain:hash

  blockchain:hash --data-path ./data
```

## `swaptoshi-core blockchain import FILEPATH`

Import from <FILE>.

```
USAGE
  $ swaptoshi-core blockchain import [FILEPATH] [-d <value>] [-f]

ARGUMENTS
  FILEPATH  Path to the gzipped blockchain data.

FLAGS
  -d, --data-path=<value>  Directory path to specify where node data is stored. Environment variable "KLAYR_DATA_PATH"
                           can also be used.
  -f, --force              Delete and overwrite existing blockchain data

DESCRIPTION
  Import from <FILE>.

EXAMPLES
  blockchain:import ./path/to/blockchain.tar.gz

  blockchain:import ./path/to/blockchain.tar.gz --data-path ./klayr/

  blockchain:import ./path/to/blockchain.tar.gz --data-path ./klayr/ --force
```

## `swaptoshi-core blockchain reset`

Reset the blockchain data.

```
USAGE
  $ swaptoshi-core blockchain reset [-d <value>] [-y]

FLAGS
  -d, --data-path=<value>  Directory path to specify where node data is stored. Environment variable "KLAYR_DATA_PATH"
                           can also be used.
  -y, --yes                Skip confirmation prompt.

DESCRIPTION
  Reset the blockchain data.

EXAMPLES
  blockchain:reset

  blockchain:reset --data-path ./klayr

  blockchain:reset --yes
```
