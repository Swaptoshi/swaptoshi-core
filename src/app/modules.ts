/* eslint-disable @typescript-eslint/no-empty-function */
import { Application, Modules } from 'klayr-sdk';
import { Modules as SwaptoshiModules } from 'swaptoshi-sdk';

interface KlayrMethod {
	validator: Modules.Validators.ValidatorsMethod;
	auth: Modules.Auth.AuthMethod;
	token: Modules.Token.TokenMethod;
	fee: Modules.Fee.FeeMethod;
	random: Modules.Random.RandomMethod;
	reward: Modules.DynamicReward.DynamicRewardMethod;
	pos: Modules.PoS.PoSMethod;
	interoperability: Modules.Interoperability.SidechainInteroperabilityMethod | Modules.Interoperability.MainchainInteroperabilityMethod;
}

export const registerModules = (app: Application, method: KlayrMethod): void => {
	const nftModule = new Modules.NFT.NFTModule();
	const tokenFactoryModule = new SwaptoshiModules.TokenFactory.TokenFactoryModule();
	const dexModule = new SwaptoshiModules.DEX.DexModule();
	const feeConversionModule = new SwaptoshiModules.FeeConversion.FeeConversionModule();
	const liquidPosModule = new SwaptoshiModules.LiquidPoS.LiquidPosModule();
	const governanceModule = new SwaptoshiModules.Governance.GovernanceModule();

	nftModule.addDependencies(method.interoperability, method.fee, method.token);
	governanceModule.addDependencies({ tokenMethod: method.token, feeMethod: method.fee });
	liquidPosModule.addDependencies({ tokenMethod: method.token, governanceMethod: governanceModule.method });
	feeConversionModule.addDependencies({ tokenMethod: method.token, feeMethod: method.fee, dexMethod: dexModule.method, governanceMethod: governanceModule.method });
	dexModule.addDependencies({
		tokenMethod: method.token,
		nftMethod: nftModule.method,
		feeMethod: method.fee,
		interoperabilityMethod: method.interoperability,
		feeConversionMethod: feeConversionModule.method,
		governanceMethod: governanceModule.method,
	});
	tokenFactoryModule.addDependencies({
		tokenMethod: method.token,
		feeMethod: method.fee,
		nftMethod: nftModule.method,
		interoperabilityMethod: method.interoperability,
		dexMethod: dexModule.method,
		feeConversionMethod: feeConversionModule.method,
		governanceMethod: governanceModule.method,
	});

	// governance module needs to executed before dynamicReward module for treasury blockRewardTax
	app.registerModulePriority(governanceModule);

	// feeConversion module needs to be executed before fee module for altering fee verification, and before all modules that implements fc_method
	app.registerModulePriority(feeConversionModule);

	app.registerModule(nftModule);
	app.registerModule(tokenFactoryModule);
	app.registerModule(dexModule);
	app.registerModule(liquidPosModule);

	// above registration order will result in following module array:
	// [feeConversion, governance, ...klayrModules, nftModule, tokenFactoryModule, dexModule, liquidPosModule]
	// the later the registerModulePriority, the higher the priority (unshift)

	app.registerInteroperableModule(nftModule);
	app.registerInteroperableModule(tokenFactoryModule);
	app.registerInteroperableModule(dexModule);
};
