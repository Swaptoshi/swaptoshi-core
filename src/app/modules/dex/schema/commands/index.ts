export * from './burn_command';
export * from './collect_command';
export * from './collect_treasury_command';
export * from './create_pool_command';
export * from './decrease_liquidity_command';
export * from './exact_input_command';
export * from './exact_input_single_command';
export * from './exact_output_command';
export * from './exact_output_single_command';
export * from './increase_liquidity_command';
export * from './mint_command';
export * from './treasurify_command';
