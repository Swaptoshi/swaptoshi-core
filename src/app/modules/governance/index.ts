export { GovernanceModule } from './module';
export { GovernanceMethod } from './method';
export { BaseGovernableConfig } from './base_governable_config';
export * from './types';
