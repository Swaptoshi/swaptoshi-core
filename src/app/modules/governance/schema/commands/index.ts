export * from './create_proposal_command';
export * from './vote_command';
export * from './boost_vote_command';
export * from './delegate_vote_command';
export * from './revoke_delegated_vote_command';
export * from './set_proposal_attributes';
