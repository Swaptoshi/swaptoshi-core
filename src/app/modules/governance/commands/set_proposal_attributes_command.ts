/* eslint-disable */
import { BaseCommand, CommandVerifyContext, CommandExecuteContext, VerificationResult, VerifyStatus } from 'klayr-sdk';
import { SetProposalAttributesParams } from '../types';
import { immutableTransactionHookGovernanceContext, mutableTransactionHookGovernanceContext } from '../stores/context';
import { ProposalStore } from '../stores/proposal';
import { setProposalAttributesCommandSchema } from '../schema';

export class SetProposalAttributesCommand extends BaseCommand {
	public async verify(_context: CommandVerifyContext<SetProposalAttributesParams>): Promise<VerificationResult> {
		try {
			const context = immutableTransactionHookGovernanceContext(_context);
			const proposalStore = await this.stores.get(ProposalStore).getImmutableProposal(context, _context.params.proposalId);
			await proposalStore.verifySetAttributes(_context.params);
		} catch (error: unknown) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error((error as { message: string }).message),
			};
		}
		return { status: VerifyStatus.OK };
	}

	public async execute(_context: CommandExecuteContext<SetProposalAttributesParams>): Promise<void> {
		const context = mutableTransactionHookGovernanceContext(_context);
		const proposalStore = await this.stores.get(ProposalStore).getMutableProposal(context, _context.params.proposalId);
		await proposalStore.setAttributes(_context.params, false);
	}

	public schema = setProposalAttributesCommandSchema;
}
