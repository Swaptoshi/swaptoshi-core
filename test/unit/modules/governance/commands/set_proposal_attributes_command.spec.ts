/* eslint-disable */
import { GovernanceModule } from '../../../../../src/app/modules/governance';
import { SetProposalAttributesCommand } from '../../../../../src/app/modules/governance/commands/set_proposal_attributes_command';

describe('SetProposalAttributesCommand', () => {
	let module: GovernanceModule;
	let command: SetProposalAttributesCommand;

	beforeEach(() => {
		module = new GovernanceModule();
		command = new SetProposalAttributesCommand(module.stores, module.events);
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toEqual('setProposalAttributes');
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
			it.todo('should throw errors for invalid schema');
			it.todo('should be ok for valid schema');
		});
	});

	describe('execute', () => {
		describe('valid cases', () => {
			it.todo('should update the state store');
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
