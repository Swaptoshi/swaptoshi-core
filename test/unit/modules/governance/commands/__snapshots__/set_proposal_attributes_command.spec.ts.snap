// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetProposalAttributesCommand constructor should have valid schema 1`] = `
{
  "$id": "/governance/command/set_proposal_attributes",
  "properties": {
    "data": {
      "dataType": "bytes",
      "fieldNumber": 3,
    },
    "key": {
      "dataType": "string",
      "fieldNumber": 2,
    },
    "proposalId": {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
  },
  "required": [
    "proposalId",
    "key",
    "data",
  ],
  "type": "object",
}
`;
